### PR TITLE
To avoid sqlite3.ProgrammingError when interleaving threads

### PR DESCRIPTION
**Pull request by chan150**
When using multi-threads application (e.g., web server gate interface; wsgi), each thread attempt to access client object which has sqlite connection.
In this case, current version of Qdrant occurs sqlite3.ProgrammingError due to the simultaneous access.

Error messages is here:
```
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread.
```

This error should be solved when your program runs on web server application (e.g., wgsi)

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you installed `pre-commit` with `pip3 install pre-commit` and set up hooks with `pre-commit install`?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: qdrant_client/local/persistence.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ def try_migrate_to_sqlite(location: str) -> None:
     try:
         dbm_storage = dbm.open(str(dbm_path), "c")
 
-        con = sqlite3.connect(str(sql_path))
+        con = sqlite3.connect(str(sql_path), , check_same_thread=False)
```
Comment (generall) [inline qdrant_client/local/persistence.py:28]: Probably a typo?

Comment (chan150) [reply qdrant_client/local/persistence.py:28]: Yes, it is a typo. 
comma with a blank should be removed.

I missed a commit before merging.

**Comment by generall** [comment]
@chan150 do you know how safe this option is in terms of data consistency? Why is it disabled by default?

**Comment by chan150** [comment]
> @chan150 do you know how safe this option is in terms of data consistency? Why is it disabled by default?

This option is safe within my knowledge.
In general, a compile macro, SQLITE_THREADSAFE, is set to 1 by default, and it make all transactions serialized.

**Comment by pieroit** [comment]
We also made some tests with concurrent websocket conversations, all of them writing in the file-based db. All is right!

+1 for merge

**Comment by imartinez** [comment]
@generall I've tested the new version 1.6.5 containing the changes in https://github.com/qdrant/qdrant-client/pull/369
Unfortunately it still doesn't work on OSX

Looking at the code in that merged branch, multi-thread is still disabled on OSX because check_same_thread is passed as True following the THREADSAFE variable.

I'd suggest to enable it by default (as proposed in this PR) or to make it optional by exposing it in the constructor (up to the client).

Note: This is preventing adoption in PrivateGPT project local setup.

**Comment by generall** [comment]
> make it optional by exposing it in the constructor (up to the client).

I think we can do this. I am still not sure enough about safety to enable it by default

**Comment by generall** [comment]
@imartinez pls check v1.6.6

**Comment by imartinez** [comment]
> @imartinez pls check v1.6.6

Hey @generall, thanks for the solution, I agree it is the right approach. 
I'm afraid it still doesn't fully work. The reason, I think, is that you forgot to pass the new param to the LocalClient here:

https://github.com/qdrant/qdrant-client/blob/f1faca44f3c78160a8500e4db8cc82f83c187aa9/qdrant_client/local/qdrant_local.py#L75

Let me know if I can help.

**Comment by generall** [comment]
@imartinez, you are right, missed that, thanks! Will fix in a sec

**Comment by generall** [comment]
@imartinez v1.6.7 :pray:

**Comment by generall** [comment]
v1.6.8

**Comment by imartinez** [comment]
@generall still fails. I've checked the commit on 1.6.8 and I don't fully get it: instead of passing the new parameter in qdrant_local:L75 (as I pointed in my prev comment), you've removed it from async_qdrant_local. Why? As far as I understand, you should pass the new param in in both cases.  

https://github.com/qdrant/qdrant-client/commit/a3748acf46f886586c2fb49c6b948c4deb69a3e6

**Comment by generall** [comment]
@imartinez I am sorry, I am just so un-focused today =(

**Comment by generall** [comment]
I hope now it is fixed

**Comment by imartinez** [comment]
It now works with 1.69. Thanks @generall. This PR can be closed because the suggestion here is already implemented.

**Comment by joein** [comment]
@chan150 thanks for the initial proposal, we've added support for this functionality in 1.6.9

**Pull request closed without merging**